api/src/main/java/org/openmrs/api/db/hibernate/HibernateOrderDAO.java

```diff
@@ -183,6 +183,35 @@ public Order getOrder(Integer orderId) throws DAOException {
 			cal.setTime(searchCriteria.getActivatedOnOrAfterDate());
 			crit.add(Restrictions.ge("dateActivated", OpenmrsUtil.firstSecondOfDay(cal.getTime())));
 		}
+		if (searchCriteria.getDateStoppedOnOrBeforeDate() != null) {
+			// an order is considered Canceled regardless of the time when the dateStopped was set
+			crit.add(Restrictions.isNotNull("dateStopped"));
+		}
+		if (searchCriteria.getAutoExpireOnOrBeforeDate() != null) {
+			// set the date's time to the last millisecond of the date
+			Calendar cal = Calendar.getInstance();
+			cal.setTime(searchCriteria.getAutoExpireOnOrBeforeDate());
+			crit.add(Restrictions.le("autoExpireDate", OpenmrsUtil.getLastMomentOfDay(cal.getTime())));
+		}
+        if (searchCriteria.getAction() != null) {
+            crit.add(Restrictions.eq("action", searchCriteria.getAction()));
+        }
+		if (searchCriteria.getFulfillerStatus() != null) {
+			crit.add(Restrictions.eq("fulfillerStatus", searchCriteria.getFulfillerStatus()));
+		}
+		if (searchCriteria.getExcludeCanceledAndExpired() == true) {
+			Calendar cal = Calendar.getInstance();
+			// exclude expired orders (include only orders with autoExpireDate = null or autoExpireDate in the future)
+			crit.add(Restrictions.or(Restrictions.isNull("autoExpireDate"), Restrictions.gt("autoExpireDate", OpenmrsUtil.getLastMomentOfDay(cal.getTime()))));
+			// exclude Canceled Orders
+			crit.add(Restrictions.isNull("dateStopped"));
```

Comment:
We should make this logic the same as the autoExpireDate.  I would assume that dateStopped shouldn't be in the future, but unfortunately the OrderValidator doesn't seem to enforce this... in the case that someone set a date stopped in the future, we wouldn't want to exclude this order.

Reply:
no problem, I will change this line to:
`crit.add(Restrictions.le("dateStopped", cal.getTime()));`

Reply:
Actually, after reading the next comment, maybe this line should be?
`crit.add(Restrictions.le("dateStopped", OpenmrsUtil.getLastMomentOfDay(cal.getTime())));`

Reply:
third time's the charm:
`// exclude Canceled Orders
			crit.add(Restrictions.or(
					Restrictions.isNull("dateStopped"),
					Restrictions.gt("dateStopped", cal.getTime())));`

Reply:
yep, that looks right...